File: app/src/main/java/com/legue/axel/lolsummonertool/utils/MasteryUtils.java
package com.legue.axel.lolsummonertool.utils;

import com.legue.axel.lolsummonertool.AppExecutors;
import com.legue.axel.lolsummonertool.database.SummonerToolDatabase;
import com.legue.axel.lolsummonertool.database.model.mastery.Mastery;
import com.legue.axel.lolsummonertool.database.model.mastery.MasteryImage;
import com.legue.axel.lolsummonertool.network.response.mastery.MasteryDetailResponse;
import com.legue.axel.lolsummonertool.network.response.mastery.MasteryResponse;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.List;

public class MasteryUtils {

    private static List<Mastery> masteries;
    private static List<MasteryImage> masteryImages;

    public static void insertMasteryResponseInDB(MasteryResponse masteryResponse, SummonerToolDatabase database) {
        masteries = new ArrayList<>();
        masteryImages = new ArrayList<>();


        if (masteryResponse != null) {
            extractMasteryDetails(masteryResponse.getMasteryDetailResponse());

            AppExecutors.Companion.getInstance().getDiskIO().execute(() -> {
                try {
                    //TODO : find a way to avoid this delete every time
                    database.masteryDao().deleteAll();
                    database.masteryImageDao().deleteAll();

                    database.masteryDao().insertAllMasteries(masteries);
                    database.masteryImageDao().insertAllMasteryImages(masteryImages);

                } catch (Exception e) {
                    e.printStackTrace();
                }

            });
        }
    }

    private static void extractMasteryDetails(LinkedHashMap<String, MasteryDetailResponse> responseLinkedHashMap) {

        Iterator it = responseLinkedHashMap.entrySet().iterator();
        while (it.hasNext()) {
            HashMap.Entry pair = (HashMap.Entry) it.next();
            String key = (String) pair.getKey();
            extractMastery((MasteryDetailResponse) pair.getValue());
            it.remove();
        }
    }

    private static void extractMastery(MasteryDetailResponse masteryDetailResponse) {
        Mastery mastery = new Mastery(
                masteryDetailResponse.getId(),
                masteryDetailResponse.getName(),
                masteryDetailResponse.getDescription(),
                masteryDetailResponse.getRanks(),
                masteryDetailResponse.getPrereq()
        );
        masteries.add(mastery);


        MasteryImage masteryImage = new MasteryImage(
                null,
                masteryDetailResponse.getImage().getFull(),
                masteryDetailResponse.getImage().getGroup(),
                masteryDetailResponse.getImage().getSprite(),
                masteryDetailResponse.getImage().getX(),
                masteryDetailResponse.getImage().getY(),
                masteryDetailResponse.getImage().getH(),
                masteryDetailResponse.getImage().getW(),
                masteryDetailResponse.getId()
        );
        masteryImages.add(masteryImage);


    }
}

File: app/src/main/java/com/legue/axel/lolsummonertool/utils/MasteryUtils.kt
package com.legue.axel.lolsummonertool.utils

import com.legue.axel.lolsummonertool.AppExecutors.Companion.instance
import com.legue.axel.lolsummonertool.database.SummonerToolDatabase
import com.legue.axel.lolsummonertool.database.model.mastery.Mastery
import com.legue.axel.lolsummonertool.database.model.mastery.MasteryImage
import com.legue.axel.lolsummonertool.network.response.mastery.MasteryDetailResponse
import com.legue.axel.lolsummonertool.network.response.mastery.MasteryResponse
import java.util.*

object MasteryUtils {
    private lateinit var masteries: MutableList<Mastery>
    private lateinit var masteryImages: MutableList<MasteryImage>
    fun insertMasteryResponseInDB(masteryResponse: MasteryResponse?, database: SummonerToolDatabase) {
        masteries = arrayListOf()
        masteryImages = arrayListOf()
        if (masteryResponse != null) {
            extractMasteryDetails(masteryResponse.masteryDetailResponse)
            instance.diskIO.execute {
                try { //TODO : find a way to avoid this delete every time
                    database.masteryDao().deleteAll()
                    database.masteryImageDao().deleteAll()
                    database.masteryDao().insertAllMasteries(masteries)
                    database.masteryImageDao().insertAllMasteryImages(masteryImages)
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }
    }

    private fun extractMasteryDetails(responseLinkedHashMap: LinkedHashMap<String, MasteryDetailResponse>) {
        responseLinkedHashMap.entries.forEach { (index, value) ->
            run {
                extractMastery(value)
            }
        }
    }

    private fun extractMastery(masteryDetailResponse: MasteryDetailResponse) {
        val mastery = Mastery(
                masteryDetailResponse.id,
                masteryDetailResponse.name,
                masteryDetailResponse.description,
                masteryDetailResponse.ranks,
                masteryDetailResponse.prereq
        )
        masteries.add(mastery)
        val masteryImage = MasteryImage(
                null,
                masteryDetailResponse.image.full,
                masteryDetailResponse.image.group,
                masteryDetailResponse.image.sprite,
                masteryDetailResponse.image.x,
                masteryDetailResponse.image.y,
                masteryDetailResponse.image.h,
                masteryDetailResponse.image.w,
                masteryDetailResponse.id
        )
        masteryImages.add(masteryImage)
    }
}